File: src/algorithms/trees/construction/serialize-deserialize-tree/sources/serialize-deserialize-tree_test.rs
include!("serialize-deserialize-tree.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<TreeNode>>, right: Option<Box<TreeNode>>) -> Option<Box<TreeNode>> {
        Some(Box::new(TreeNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<TreeNode>> {
        make_node(value, None, None)
    }

    fn inorder(root: &Option<Box<TreeNode>>) -> Vec<i32> {
        match root {
            None => vec![],
            Some(node) => {
                let mut result = inorder(&node.left);
                result.push(node.value);
                result.extend(inorder(&node.right));
                result
            }
        }
    }

    #[test]
    fn test_serializes_null_as_null_string() {
        assert_eq!(serialize_tree(None), "null");
    }

    #[test]
    fn test_deserializes_null_string() {
        assert!(deserialize_tree("null").is_none());
    }

    #[test]
    fn test_round_trips_balanced_7_node_bst() {
        let original = make_node(4,
            make_node(2, leaf(1), leaf(3)),
            make_node(6, leaf(5), leaf(7)));
        let serialized = serialize_tree(original.clone().map(|n| *n));
        let reconstructed = deserialize_tree(&serialized);
        assert_eq!(reconstructed.as_ref().unwrap().value, 4);
        assert_eq!(inorder(&reconstructed), vec![1, 2, 3, 4, 5, 6, 7]);
    }

    #[test]
    fn test_round_trips_single_node() {
        let original = leaf(99);
        let serialized = serialize_tree(original.clone().map(|n| *n));
        let reconstructed = deserialize_tree(&serialized);
        assert_eq!(reconstructed.as_ref().unwrap().value, 99);
        assert!(reconstructed.as_ref().unwrap().left.is_none());
        assert!(reconstructed.as_ref().unwrap().right.is_none());
    }
}

File: src/algorithms/trees/construction/serialize-deserialize-tree/serialize-deserialize-tree_test.go
package main

import (
	"reflect"
	"testing"
)

func makeSDTNode(value int, left *TreeNode, right *TreeNode) *TreeNode {
	return &TreeNode{value: value, left: left, right: right}
}

func sdtLeaf(value int) *TreeNode {
	return &TreeNode{value: value}
}

func sdtInorder(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	left := sdtInorder(root.left)
	right := sdtInorder(root.right)
	result := append(left, root.value)
	return append(result, right...)
}

func TestSerializeTreeNull(t *testing.T) {
	if serializeTree(nil) != "null" {
		t.Error("null should serialize to 'null'")
	}
}

func TestDeserializeTreeNull(t *testing.T) {
	if deserializeTree("null") != nil {
		t.Error("null string should deserialize to nil")
	}
}

func TestRoundTripsBalanced7NodeBST(t *testing.T) {
	original := makeSDTNode(4,
		makeSDTNode(2, sdtLeaf(1), sdtLeaf(3)),
		makeSDTNode(6, sdtLeaf(5), sdtLeaf(7)))
	serialized := serializeTree(original)
	reconstructed := deserializeTree(serialized)
	if reconstructed == nil || reconstructed.value != 4 {
		t.Error("reconstructed root should be 4")
	}
	if !reflect.DeepEqual(sdtInorder(reconstructed), []int{1, 2, 3, 4, 5, 6, 7}) {
		t.Error("inorder should match")
	}
}

func TestRoundTripsSingleNode(t *testing.T) {
	original := sdtLeaf(99)
	serialized := serializeTree(original)
	reconstructed := deserializeTree(serialized)
	if reconstructed == nil || reconstructed.value != 99 {
		t.Error("single node value should be 99")
	}
	if reconstructed.left != nil || reconstructed.right != nil {
		t.Error("single node should have no children")
	}
}
